File: agent/models/storyboard_landbank_v3_round2.py
"""Strict Round 2 envelopes for the V3 Copy Register and AI assistant.

These models are deliberately separate from the V2 copy authority and from the
Round 1 deterministic records.  Provider output is accepted only through the
``V3AIProviderEnvelope`` contract; approval and run receipts are immutable
provenance envelopes, never executable V2/P6 payloads.
"""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from agent.models.storyboard_landbank_v3 import (
    V3BridgeContract,
    V3RevisionRef,
    V3SemanticClass,
    deterministic_digest,
    digest_text,
    normalized_text,
)


_SHA256 = r"^[0-9a-f]{64}$"
AssistantMode = Literal["CREATE", "EXPAND", "FILL_CAPACITY"]
ProviderMode = Literal["LIVE_TEXT_ASSIST", "FAKE_TEST"]
ApprovalScope = Literal["INDIVIDUAL", "BATCH"]
ApprovalTarget = Literal["MASTER_STORYBOARD", "DURATION_PROJECTION"]


class V3ProviderSummary(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    lane: Literal["text_assist"] = "text_assist"
    status: str = Field(min_length=1)
    configured: bool = False
    provider_id: str | None = None
    model_id: str | None = None
    execution_enabled: bool = False
    provider_calls: int = Field(default=0, ge=0)
    fake_provider_allowed: bool = False


class V3AssistantGap(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    semantic_class: Literal["HOOK", "BODY_CORE", "CTA"]
    current_count: int = Field(ge=0, le=500)
    target_count: int = Field(ge=0, le=500)
    gap_count: int = Field(ge=0, le=24)
    reason: str = Field(min_length=1, max_length=240)


class V3AssistantPlan(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["3.2"] = "3.2"
    plan_id: str = Field(min_length=1)
    run_id: str = Field(min_length=1)
    product_id: str = Field(min_length=1)
    recipe: V3RevisionRef
    # The plan is a durable authoring contract, not only a count calculation.
    # Keep the locked route and truth snapshot beside the run so a reviewer can
    # reconstruct what the assistant was allowed to see and produce.
    objective: dict[str, Any] = Field(default_factory=dict)
    formula: dict[str, Any] = Field(default_factory=dict)
    angle: V3RevisionRef | None = None
    storyline_family: V3RevisionRef | None = None
    product_truth: dict[str, Any] = Field(default_factory=dict)
    evidence_fact_ids: tuple[str, ...] = Field(default_factory=tuple, max_length=500)
    evidence_digest: str = Field(default="0" * 64, pattern=_SHA256)
    language_profile: str = Field(default="Malay", min_length=1)
    current_capacity: dict[str, int] = Field(default_factory=dict)
    mode: AssistantMode
    target_counts: dict[str, int] = Field(default_factory=dict)
    gaps: tuple[V3AssistantGap, ...] = Field(default_factory=tuple)
    target_durations_seconds: tuple[int, ...] = Field(min_length=1, max_length=3)
    wps_mode: Literal["SAFE", "SWEET"]
    provider: V3ProviderSummary
    prompt_version: str = Field(min_length=1)
    prompt_digest: str = Field(pattern=_SHA256)
    estimated_provider_calls: int = Field(ge=0, le=3)
    estimated_output_tokens: int = Field(ge=0, le=20000)
    estimated_credit_spend: int = Field(default=0, ge=0)
    max_proposals: int = Field(ge=1, le=24)
    max_provider_calls: int = Field(default=1, ge=0, le=1)
    max_output_tokens: int = Field(default=20000, ge=1, le=20000)
    max_cost: int = Field(default=0, ge=0)
    cost_status: Literal["NOT_REPORTED", "WITHIN_BUDGET", "BUDGET_EXCEEDED"] = "NOT_REPORTED"
    explicit_execute_required: Literal[True] = True
    created_at: str = Field(min_length=1)
    created_by: str = Field(min_length=1)


class V3AICopySegment(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    formula_stage_key: str = Field(min_length=1, max_length=80)
    authored_text: str = Field(min_length=8, max_length=1200)
    entry_key: str = Field(min_length=1, max_length=120)
    exit_key: str = Field(min_length=1, max_length=120)
    continuity_requirements: tuple[str, ...] = Field(default_factory=tuple, max_length=8)
    evidence_fact_ids: tuple[str, ...] = Field(default_factory=tuple, max_length=12)
    claim_bearing: bool = True

    @field_validator("authored_text", "entry_key", "exit_key")
    @classmethod
    def _clean_text(cls, value: str) -> str:
        value = normalized_text(value)
        if not value:
            raise ValueError("V3_AI_TEXT_REQUIRED")
        return value


class V3AICopyProposal(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    proposal_id: str = Field(min_length=1, max_length=120)
    semantic_class: Literal["HOOK", "BODY_CORE", "CTA"]
    angle_definition: str = Field(min_length=8, max_length=500)
    storyline_definition: str = Field(min_length=8, max_length=500)
    segments: tuple[V3AICopySegment, ...] = Field(min_length=1, max_length=8)
    rationale: str = Field(min_length=8, max_length=800)
    risk_notes: tuple[str, ...] = Field(default_factory=tuple, max_length=12)


class V3AIProviderEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["v3-copy-assistant-1"] = "v3-copy-assistant-1"
    proposals: tuple[V3AICopyProposal, ...] = Field(min_length=1, max_length=24)


class V3AIProviderReceipt(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    mode: ProviderMode
    lane: Literal["text_assist"] = "text_assist"
    provider_id: str | None = None
    model_id: str | None = None
    call_id: int | None = Field(default=None, ge=0)
    response_status: str = Field(min_length=1)
    json_parse_status: str = Field(min_length=1)
    usage: dict[str, int | float] = Field(default_factory=dict)
    prompt_digest: str = Field(pattern=_SHA256)
    output_digest: str = Field(pattern=_SHA256)


class V3QualitySignal(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    hard_pass: bool
    formula_valid: bool
    evidence_valid: bool
    bridge_valid: bool
    claim_safety_valid: bool
    truth_current: bool
    wps_valid: bool
    issue_codes: tuple[str, ...] = Field(default_factory=tuple)
    novelty_signal: Literal["NOVEL", "NEAR_DUPLICATE", "EXACT_DUPLICATE", "UNKNOWN"]
    novelty_score: float = Field(ge=0.0, le=1.0)
    quality_score: float = Field(ge=0.0, le=1.0)


class V3AssistantRunReceipt(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    run_id: str = Field(min_length=1)
    plan_id: str = Field(min_length=1)
    product_id: str = Field(min_length=1)
    recipe: V3RevisionRef
    mode: AssistantMode
    status: Literal["PLANNED", "EXECUTED", "FAILED"]
    provider: V3AIProviderReceipt | None = None
    proposal_ids: tuple[str, ...] = Field(default_factory=tuple)
    component_refs: tuple[V3RevisionRef, ...] = Field(default_factory=tuple)
    master: V3RevisionRef | None = None
    projections: tuple[V3RevisionRef, ...] = Field(default_factory=tuple)
    output_digest: str | None = Field(default=None, pattern=_SHA256)
    provider_calls: int = Field(default=0, ge=0)
    credit_spend: int = Field(default=0, ge=0)
    token_usage: dict[str, int | float] = Field(default_factory=dict)
    quality: V3QualitySignal | None = None
    error_code: str | None = None
    created_at: str = Field(min_length=1)
    created_by: str = Field(min_length=1)


class V3ApprovalChecklist(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    semantic_reviewed: bool = False
    product_truth_reviewed: bool = False
    formula_reviewed: bool = False
    evidence_reviewed: bool = False
    bridge_reviewed: bool = False
    safety_reviewed: bool = False
    duration_reviewed: bool = False

    def all_passed(self) -> bool:
        return all(self.model_dump(mode="json").values())


class V3HumanApprovalReceipt(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["v3-human-approval-receipt-1"] = "v3-human-approval-receipt-1"
    receipt_id: str = Field(min_length=1)
    approval_scope: ApprovalScope
    target_type: ApprovalTarget
    target_id: str = Field(min_length=1)
    target_revision: int = Field(ge=1)
    product_id: str = Field(min_length=1)
    master_ref: V3RevisionRef
    projection_refs: tuple[V3RevisionRef, ...] = Field(default_factory=tuple, max_length=3)
    batch_target_refs: tuple[V3RevisionRef, ...] = Field(default_factory=tuple, max_length=24)
    exact_content_fingerprint: str = Field(pattern=_SHA256)
    projection_fingerprints: tuple[str, ...] = Field(default_factory=tuple, max_length=3)
    product_truth_snapshot_id: str = Field(min_length=1)
    product_truth_snapshot_version: int = Field(ge=1)
    product_truth_snapshot_digest: str = Field(pattern=_SHA256)
    formula_id: str = Field(min_length=1)
    formula_version: str = Field(min_length=1)
    evidence_digest: str = Field(pattern=_SHA256)
    wps_authority_digests: tuple[str, ...] = Field(default_factory=tuple, max_length=3)
    checklist: V3ApprovalChecklist
    approved_by: str = Field(min_length=1)
    rationale: str = Field(min_length=8, max_length=1200)
    automatic_approval: Literal[False] = False
    batch_id: str | None = None
    receipt_digest: str = Field(pattern=_SHA256)
    created_at: str = Field(min_length=1)


class V3PromptPreview(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    plan_id: str = Field(min_length=1)
    prompt_version: str = Field(min_length=1)
    prompt_digest: str = Field(pattern=_SHA256)
    system_instructions: str = Field(min_length=1)
    untrusted_truth_json: str = Field(min_length=2)
    requested_output_contract: dict[str, Any]
    provider_calls: int = 0
    credit_spend: int = 0


def approval_receipt_digest(receipt: V3HumanApprovalReceipt) -> str:
    payload = receipt.model_dump(mode="json")
    payload.pop("receipt_digest", None)
    return deterministic_digest(payload)


def proposal_text_digest(proposal: V3AICopyProposal) -> str:
    return deterministic_digest({
        "semantic_class": proposal.semantic_class,
        "segments": [
            {
                **segment.model_dump(mode="json"),
                "text_digest": digest_text(segment.authored_text),
            }
            for segment in proposal.segments
        ],
    })


__all__ = [
    "AssistantMode",
    "ProviderMode",
    "V3ProviderSummary",
    "V3AssistantGap",
    "V3AssistantPlan",
    "V3AICopySegment",
    "V3AICopyProposal",
    "V3AIProviderEnvelope",
    "V3AIProviderReceipt",
    "V3QualitySignal",
    "V3AssistantRunReceipt",
    "V3ApprovalChecklist",
    "V3HumanApprovalReceipt",
    "V3PromptPreview",
    "approval_receipt_digest",
    "proposal_text_digest",
]

File: dashboard/src/api/storyboardLandbankV3Round2.ts
import { fetchAPI, getAPI, postAPI } from "./client";

export type V3AssistantMode = "CREATE" | "EXPAND" | "FILL_CAPACITY";

export interface V3ProviderStatus {
	lane: "text_assist";
	status: string;
	configured: boolean;
	provider_id?: string | null;
	model_id?: string | null;
	execution_enabled: boolean;
	provider_calls: number;
	credit_spend: number;
	fake_provider_allowed: boolean;
}

export interface V3AssistantGap {
	semantic_class: "HOOK" | "BODY_CORE" | "CTA";
	current_count: number;
	target_count: number;
	gap_count: number;
	reason: string;
}

export interface V3AssistantPlan {
	plan_id: string;
	run_id: string;
	product_id: string;
	recipe: { entity_id: string; revision: number };
	objective?: Record<string, string>;
	formula?: { formula_id?: string; formula_version?: string };
	angle?: { entity_id: string; revision: number } | null;
	storyline_family?: { entity_id: string; revision: number } | null;
	product_truth?: Record<string, unknown>;
	evidence_fact_ids?: string[];
	evidence_digest?: string;
	language_profile?: string;
	current_capacity?: Record<string, number>;
	mode: V3AssistantMode;
	target_counts: Record<string, number>;
	gaps: V3AssistantGap[];
	target_durations_seconds: number[];
	wps_mode: "SAFE" | "SWEET";
	provider: V3ProviderStatus;
	prompt_version: string;
	prompt_digest: string;
	estimated_provider_calls: number;
	estimated_output_tokens: number;
	estimated_credit_spend: number;
	max_proposals: number;
	max_provider_calls?: number;
	max_output_tokens?: number;
	max_cost?: number;
	cost_status?: string;
	supply_actions?: Record<string, string>;
	diversity_deficits?: string[];
	marginal_plan?: Record<string, number>;
	capacity_before?: { reviewable_capacity?: number; theoretical_capacity?: number; duration_counts?: Record<string, number> };
	evidence_selection?: { outcome?: string; fact_ids?: string[]; explanations?: Record<string, string[]>; score_by_fact?: Record<string, number> };
	explicit_execute_required: true;
	created_at: string;
	created_by: string;
}

export interface V3TruthFact {
	fact_id: string;
	fact_kind: string;
	text: string;
	text_digest: string;
	approved: boolean;
}

export interface V3QualitySignal {
	hard_pass: boolean;
	formula_valid: boolean;
	evidence_valid: boolean;
	bridge_valid: boolean;
	claim_safety_valid: boolean;
	truth_current: boolean;
	wps_valid: boolean;
	issue_codes: readonly string[];
	novelty_signal: "NOVEL" | "NEAR_DUPLICATE" | "EXACT_DUPLICATE" | "UNKNOWN";
	novelty_score: number;
	quality_dimensions?: Record<string, number>;
	quality_score: number;
}

export interface V3StoryboardStage {
	stage_key: string;
	formula_stage_key: string;
	semantic_class: string;
	authored_text: string;
	entry_key?: string;
	exit_key?: string;
	evidence_fact_ids?: string[];
	claim_bearing?: boolean;
	text_digest?: string;
}

// Macro Round 3 (P3): read-only V2 production materialization status. A single
// projection is NOT_MATERIALIZED / MATERIALIZED / STALE / BLOCKED; the item rollup
// additionally reports PARTIALLY_MATERIALIZED when only some projections are live.
export type V3MaterializationStatus = "NOT_MATERIALIZED" | "MATERIALIZED" | "STALE" | "BLOCKED";
export type V3MaterializationRollup = V3MaterializationStatus | "PARTIALLY_MATERIALIZED";

export interface V3ProjectionMaterialization {
	status: V3MaterializationStatus;
	link_id?: string | null;
	v2_blueprint_id?: string;
	v2_blueprint_revision?: number;
	v2_approval_snapshot_id?: string;
	materialization_digest?: string;
	reason?: string;
}

export interface V3Projection {
	projection_id: string;
	revision: number;
	target_duration_seconds: number;
	exact_resolved_dialogue: string;
	derivation_source: string;
	authoring_run_id?: string | null;
	status: string;
	per_block_word_budgets: number[];
	wps_mode?: string;
	wps_authority_digest?: string;
	stage_allocations?: Array<{ master_stage_key: string; transform_mode: string; projected_text?: string; target_block_indices?: number[] }>;
	// Read-only enrichment from the GET landbank endpoint (Round 3 P3).
	exact_projection_digest?: string;
	materialization?: V3ProjectionMaterialization;
}

export interface V3LandbankItem {
	master: {
		master_id: string;
		revision: number;
	product_id: string;
	objective?: { objective_id?: string; definition?: string };
	recipe?: { entity_id: string; revision: number };
	formula: { formula_id: string; formula_version: string };
		angle: { entity_id: string; revision: number };
		storyline_family: { entity_id: string; revision: number };
	stages: readonly V3StoryboardStage[];
		status: string;
		source: string;
		exact_content_digest: string;
		word_count: number;
		resolved_component_refs?: Array<{ entity_id: string; revision: number }>;
	};
	projections: readonly V3Projection[];
	quality: V3QualitySignal;
	current_truth: boolean;
	approval_receipt?: V3ApprovalReceipt | null;
	// Round 3 P3 replaces the Round 2 "NOT_IN_ROUND2" stub with a truthful rollup.
	v2_materialization: V3MaterializationRollup | "NOT_IN_ROUND2";
	p6_status: string;
}

export interface V3ApprovalReceipt {
	receipt_id?: string;
	[key: string]: unknown;
}

export interface V3LandbankResponse {
	source: "V3_COPY_REGISTER";
	product_id: string;
	items: V3LandbankItem[];
	total: number;
	limit: number;
	offset: number;
	has_more: boolean;
	provider_calls: number;
	v2_mixed: false;
	full_storyboard_first: true;
}

export interface V3AssistantExecution {
	run_id: string;
	plan_id: string;
	status: "EXECUTED";
	provider: { mode: "LIVE_TEXT_ASSIST" | "FAKE_TEST"; provider_id?: string | null; model_id?: string | null };
	master: { entity_id: string; revision: number };
	projections: Array<{ entity_id: string; revision: number }>;
	quality: V3QualitySignal;
	provider_calls: number;
	credit_spend: number;
	projection_derivation: string;
}

export interface V3ApprovalChecklist {
	semantic_reviewed: boolean;
	product_truth_reviewed: boolean;
	formula_reviewed: boolean;
	evidence_reviewed: boolean;
	bridge_reviewed: boolean;
	safety_reviewed: boolean;
	duration_reviewed: boolean;
}

function requestId(prefix: string): string {
	const uuid = typeof crypto !== "undefined" && "randomUUID" in crypto ? crypto.randomUUID() : `${Date.now()}-${Math.random()}`;
	return `${prefix}:${uuid}`;
}

export type V3RecipePreset = "QUICK TEST" | "FAST54" | "MULTI-ANGLE" | "SCALE" | "CUSTOM";

export async function fetchV3CopyRegisterProviderStatus(): Promise<V3ProviderStatus> {
	return getAPI<V3ProviderStatus>("/api/storyboard-landbank/v3/copy-register/provider-status");
}

export async function setupV3Campaign(input: {
	product_id: string;
	objective_id: string;
	objective_definition?: string;
	formula_id: string;
	preset: V3RecipePreset;
	supported_durations_seconds?: number[];
	target_capacity?: number;
	language_profile?: string;
	wps_mode?: "SAFE" | "SWEET";
	component_count_targets?: Record<string, number>;
}): Promise<{ recipe_id: string; recipe_revision: number; preset: string; reused: boolean; recipe: Record<string, unknown> }> {
	return postAPI("/api/storyboard-landbank/v3/copy-register/setup-campaign", {
		...input,
		actor_id: "dashboard-operator",
		request_id: requestId("v3-setup"),
	});
}

export async function deriveV3AiAssistedProjection(input: {
	master_id: string;
	master_revision?: number;
	duration_seconds: number;
	provider_mode: "LIVE_TEXT_ASSIST" | "FAKE_TEST";
	language_profile?: string;
	wps_mode?: "SAFE" | "SWEET";
}): Promise<{ derivation_source: string; compressed_stages: string[]; projection: Record<string, unknown>; automatic_approval: false }> {
	return postAPI("/api/storyboard-landbank/v3/copy-register/assistant/projection/ai-assisted", {
		...input,
		actor_id: "dashboard-operator",
		request_id: requestId("v3-ai-projection"),
	});
}

export async function reviewV3Entity(action: "validate" | "submit" | "reject" | "archive", entityType: string, entityId: string, revision: number): Promise<Record<string, unknown>> {
	// `revision` is honoured as a query param (validate) and in the body (submit/reject/archive).
	return postAPI(`/api/storyboard-landbank/v3/review/${encodeURIComponent(entityType)}/${encodeURIComponent(entityId)}/${action}?revision=${revision}`, {
		revision,
		actor_id: "dashboard-operator",
		request_id: requestId(`v3-${action}`),
	});
}

export async function regenerateV3Component(componentId: string, revision: number, providerMode: "LIVE_TEXT_ASSIST" | "FAKE_TEST"): Promise<{ new_revision: number; source_revision: number; component: Record<string, unknown>; automatic_approval: false; run_id: string }> {
	return postAPI(`/api/storyboard-landbank/v3/copy-register/components/${encodeURIComponent(componentId)}/regenerate`, {
		revision,
		provider_mode: providerMode,
		actor_id: "dashboard-operator",
		request_id: requestId("v3-regenerate"),
	});
}

export async function deleteV3Draft(entityType: string, entityId: string, revision: number): Promise<{ deleted: boolean }> {
	// DELETE draft endpoint reads X-Actor-Id / X-Request-Id headers; backend fails closed if unsafe.
	return fetchAPI(`/api/storyboard-landbank/v3/drafts/${encodeURIComponent(entityType)}/${encodeURIComponent(entityId)}/${revision}`, {
		method: "DELETE",
		headers: { "X-Actor-Id": "dashboard-operator", "X-Request-Id": requestId("v3-delete") },
	});
}

export async function planV3Assistant(input: {
	product_id: string;
	recipe_id: string;
	mode: V3AssistantMode;
	additional_count?: number;
	semantic_class?: string;
	target_capacity?: number;
	evidence_fact_ids?: string[];
	max_provider_calls?: number;
	max_output_tokens?: number;
	max_cost?: number;
}): Promise<{ plan: V3AssistantPlan; provider_calls: number; credit_spend: number }> {
	return postAPI("/api/storyboard-landbank/v3/copy-register/assistant/plan", {
		...input,
		actor_id: "dashboard-operator",
		request_id: requestId("v3-plan"),
	});
}

export async function fetchV3AssistantPromptPreview(planId: string): Promise<{ preview: { prompt_digest: string; system_instructions: string; untrusted_truth_json: string; requested_output_contract: Record<string, unknown> }; provider_calls: number; credit_spend: number }> {
	return postAPI(`/api/storyboard-landbank/v3/copy-register/assistant/plans/${encodeURIComponent(planId)}/prompt-preview`, {});
}

export async function executeV3Assistant(planId: string, provider_mode: "LIVE_TEXT_ASSIST" | "FAKE_TEST"): Promise<V3AssistantExecution> {
	return postAPI(`/api/storyboard-landbank/v3/copy-register/assistant/plans/${encodeURIComponent(planId)}/execute`, {
		provider_mode,
		actor_id: "dashboard-operator",
		request_id: requestId("v3-execute"),
	});
}

export async function fetchV3ProductTruth(productId: string): Promise<{ product_id: string; lineage: Record<string, unknown>; facts: V3TruthFact[]; fact_count: number; provider_calls: number; mutations: number }> {
	return getAPI(`/api/storyboard-landbank/v3/products/${encodeURIComponent(productId)}/truth`);
}

export async function fetchV3CopyRegisterLandbank(productId: string, params: { status?: string; duration_seconds?: number; quality?: string; formula_id?: string; angle_id?: string; storyline_family_id?: string; recipe_id?: string; source?: string; blocker?: string; search?: string; limit?: number; offset?: number } = {}): Promise<V3LandbankResponse> {
	const query = new URLSearchParams({ product_id: productId, limit: String(params.limit ?? 50), offset: String(params.offset ?? 0) });
	if (params.status) query.set("status", params.status);
	if (params.duration_seconds) query.set("duration_seconds", String(params.duration_seconds));
	if (params.quality) query.set("quality", params.quality);
	if (params.formula_id) query.set("formula_id", params.formula_id);
	if (params.angle_id) query.set("angle_id", params.angle_id);
	if (params.storyline_family_id) query.set("storyline_family_id", params.storyline_family_id);
	if (params.recipe_id) query.set("recipe_id", params.recipe_id);
	if (params.source) query.set("source", params.source);
	if (params.blocker) query.set("blocker", params.blocker);
	if (params.search) query.set("search", params.search);
	return getAPI<V3LandbankResponse>(`/api/storyboard-landbank/v3/copy-register/landbank?${query.toString()}`);
}

export async function approveV3Master(input: {
	master_id: string;
	projection_ids: Array<{ entity_id?: string; projection_id?: string; revision: number }>;
	approved_by: string;
	rationale: string;
	checklist: V3ApprovalChecklist;
}): Promise<{ receipt: Record<string, unknown>; master: V3LandbankItem["master"]; projections: V3Projection[]; automatic_approval: false }> {
	return postAPI(`/api/storyboard-landbank/v3/copy-register/approval/master/${encodeURIComponent(input.master_id)}`, {
		...input,
		actor_id: input.approved_by,
		request_id: requestId("v3-approval"),
	});
}

export async function approveV3MasterBatch(input: {
	targets: Array<{ master_id: string; projection_ids: Array<{ entity_id?: string; projection_id?: string; revision: number }> }>;
	approved_by: string;
	rationale: string;
	checklist: V3ApprovalChecklist;
}): Promise<{ batch_id: string; batch_digest?: string; approved_count: number; automatic_approval: false }> {
	return postAPI(`/api/storyboard-landbank/v3/copy-register/approval/batch`, {
		...input,
		actor_id: input.approved_by,
		request_id: requestId("v3-batch-approval"),
	});
}

export interface V3MaterializeResult {
	projection_id: string;
	status: "MATERIALIZED";
	blueprint_status: string;
	blueprint_id: string;
	revision: number;
	link_id: string;
	v2_approval_snapshot_id: string;
	materialization_digest: string;
	idempotent_reuse: boolean;
}

export interface V3MaterializeBlocked {
	projection_id: string | null;
	code: string;
	detail: string;
	details?: Record<string, unknown> | null;
}

export interface V3MaterializeBulkResult {
	requested: number;
	materialized_count: number;
	blocked_count: number;
	materialized: V3MaterializeResult[];
	blocked: V3MaterializeBlocked[];
}

// Round 3 P3: explicitly materialize ONE approved projection into a V2
// PRODUCTION_VALID blueprint. actor_id/request_id follow the same mutation-receipt
// pattern as approveV3Master; the human approval receipt is the authority carried
// forward, so a receipt_id is mandatory server-side.
export async function materializeV3Projection(input: { productId?: string; projectionId: string; receiptId: string }): Promise<V3MaterializeResult> {
	return postAPI("/api/storyboard-landbank/v3/copy-register/materialize", {
		projection_id: input.projectionId,
		receipt_id: input.receiptId,
		...(input.productId ? { product_id: input.productId } : {}),
		actor_id: "dashboard-operator",
		request_id: requestId("v3-materialize"),
	});
}

// Round 3 P3: bounded bulk materialization (partial success). Each item is
// independent — one blocked projection never poisons the clean ones. The server
// caps the batch at 25 items per call.
export async function materializeV3ProjectionsBulk(input: { items: Array<{ projectionId: string; receiptId: string }> }): Promise<V3MaterializeBulkResult> {
	return postAPI("/api/storyboard-landbank/v3/copy-register/materialize-bulk", {
		items: input.items.map((entry) => ({ projection_id: entry.projectionId, receipt_id: entry.receiptId })),
		actor_id: "dashboard-operator",
		request_id: requestId("v3-materialize-bulk"),
	});
}
